### PR TITLE
#1433 Coupon code functionalities move to cart module API 

### PR DESCRIPTION
**Pull request by mercs600**
### Related issues

#1433 

### Short description and why it's useful
1. Core functionalities of coupon codes moved to API methods: applyCoupon, removeCoupon, appliedCoupon
2. Added unit tets
3. Moved methods responsible for UI from microcart core component to default theme
4. Updated upgrade notes
5. Updated cart API doc 

### Contribution and curently important rules acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)
- [x] I read the [TypeScript Action Plan](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/TypeScript%20Action%20Plan.md) and adjusted my PR according to it
- [x] I read about [Vue Storefront Modules](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/api-modules/about-modules.md) and [refactoring plan for them](https://github.com/DivanteLtd/vue-storefront/blob/master/doc/api-modules/refactoring-to-modules.md)

### CODE REVIEW

**Comment by mercs600** [comment]
@pkarw @filrak I will add some refactore here to clean this component to maximum.

File: core/components/blocks/Microcart/Microcart.js [hunk 1]
```diff
@@ -20,11 +20,6 @@ export default {
     VueOfflineMixin
   ],
   props: {
-    product: {
-      type: Object,
```
Comment (mercs600) [inline core/components/blocks/Microcart/Microcart.js]: @filrak I removed it. I see we don't use it anywhere. It was added at beginning -> https://github.com/DivanteLtd/vue-storefront/commit/aa0e57abd0924fb5f0f6e365deb2f220fe41c002#diff-cb47176996933f7293e2e419bbfdca34R8

**Comment by mercs600** [comment]
Now it's clean ;-)

File: core/modules/cart/features/cartPayment.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+/**
+ * Functionality for returning payment method
+ *
+ * #### Computed properties
+ * - **`cartPayment`** returns cart payment. Returns `cart/payment` Vuex state.
```
Comment (filrak) [inline core/modules/cart/features/cartPayment.ts]: It's not very descriptive. Maybe we should specify that it returns default cart payment method?

File: core/modules/cart/features/cartShipping.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+/**
+ * Functionality for returning shipping method
+ *
+ * #### Computed properties
+ * - **`cartShipping`** returns cart shipping. Returns `cart/shipping` Vuex state.
```
Comment (filrak) [inline core/modules/cart/features/cartShipping.ts]: Also not very descriptive

File: core/modules/cart/features/applyCoupon.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+/**
+ * Functionality for adding discount code.
+ *
+ * #### Methods
+ * - **`applyCoupon(code)`** adds passed coupon code. Dispatches `cart/applyCoupon` Vuex action
```
Comment (filrak) [inline core/modules/cart/features/applyCoupon.ts]: Maybe something like 'applies coupon with given code'?

File: core/modules/cart/features/appliedCoupon.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
+/**
+ * Functionality for returning provided coupon code and discount value
+ *
+ * #### Computed properties
+ * - **`appliedCoupon`** returns coupon code and provided discount.
```
Comment (filrak) [inline core/modules/cart/features/appliedCoupon.ts]: Maybe 'return currently applied coupon code and associated discount' ?

File: core/modules/cart/features/cartTotals.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Functionality for returning total segments array.
+ *
+ * #### Computed properties
+ * - **`cartTotals`** returns cart totals array. Returns `cart/totals` Vuex getter.
```
Comment (filrak) [inline core/modules/cart/features/cartTotals.ts]: We should a least tell what cart totals are

File: core/store/modules/cart/actions.ts [hunk 7]
```diff
@@ -416,46 +416,50 @@ const actions: ActionTree<CartState, RootState> = {
...
-      }, { root: true }).then(task => {
-        if (task.result) {
-          context.dispatch('refreshTotals')
-        }
-      }).catch(e => {
-        console.error(e)
-      })
-    }
+    return new Promise((resolve, reject) => {
+      if (rootStore.state.config.cart.synchronize_totals && (typeof navigator !== 'undefined' ? navigator.onLine : true)) {
```
Comment (filrak) [inline core/store/modules/cart/actions.ts]: I think we should ommit this online check. Then promis will just be rejected and we can handle error properly

**Comment by vue-kacper** [comment]
@pkarw This PR looks ok for me, mainly refactoring. It should not generate problems in future after merge with master branch

**Comment by pkarw** [comment]
@mercs600 could You please fix the comments pointed out by @filrak ?

**Comment by mercs600** [comment]
@filrak @pkarw done

**Pull request merged**